src/Shape.cpp

```diff
@@ -71,7 +71,7 @@ ShapePtr Shape::read(std::istream& input) {
     input.read((char*)&p_size, sizeof(p_size));
 
     for(int i = 0; i < p_size; ++i) {
-        float x, y, z;
+        double x, y, z;
```

Comment:
I know, but later on it is assigned to a double. So you are losing precision. Because you are going from double to float to double, it would be better to go from double to double to double.